[Performance Hints] Add simple check for returning of values of array and dictionary type

artemcm opened the pull request:
- Add simple check for returning of values of array and dictionary type
This check will run on each type-checked primary input of the current compilation and emit a warning diagnostic for all discovered occurrences of this code pattern when the performance hints feature is enabled.

- Add `@performanceOverride` attribute to silence individual performance hint diagnostics
This attribute will accept two parameters: a diagnostic group identifier first, and a string literal 'explanation' for the override reason second

aviralg reviewed (commented):
`FunctionReturnType` is too generic when we are only checking for arrays and dictionaries. We will implement other kind of return type checks such as existential any that will come under a separate category.  We should rename `FunctionReturnType` to something like `FunctionReturnsHeapAllocatedData`. This will change the name of the check, corresponding diagnostic override, and the name of the documentation file.

artemcm committed:
[Performance Hints] Add simple check for returning of values of array and dictionary type

This check will run on each type-checked primary input of the current compilation and emit a warning diagnostic for all discovered occurences of this code pattern when the performance hint diagnostic is enabled

aviralg reviewed (approved):
LGTM!

DougGregor commented on include/swift/AST/DiagnosticGroups.def:
```diff
@@ -78,6 +78,7 @@ GROUP(TrailingClosureMatching, "trailing-closure-matching")
 GROUP(UnknownWarningGroup, "unknown-warning-group")
 GROUP(CompilationCaching, "compilation-caching")
 GROUP(WeakMutability, "weak-mutability")
+GROUP(PerfHintReturnTypeImplicitCopy, "perf-hint-return-type-implicit-copy")
```
This is... very specific. Can we perhaps drop the "PerfHint"/"perf-hint-" prefix here and put them within a "PerformanceHints" group?

artemcm replied:
The goal here was to allow each check/hint to be enabled/disabled _individually_. And the idea is the each hint will actually have a small handful of diagnostic varieties, functions and closures being separate diagnostic entires in this PR being an example.

DougGregor commented on include/swift/AST/Attr.h:
```diff
@@ -730,6 +731,63 @@ class SectionAttr : public DeclAttribute {
 };
 
+/// Defines the @performanceOverride attribute.
+class PerformanceOverrideAttr : public DeclAttribute {
+public:
+  enum class CheckKind : uint8_t { ReturnTypeImplicitCopy, Count = 1 };
```
I'm not really a fan of the design here that uses an attribute to silence warnings. IMO, either there's a general feature to silence warnings within a region of the source, or we don't do it.

artemcm replied:
Removed the attribute from this PR to consider a more general design. I'd like to proceed with just the new diagnostic logic for now.

DougGregor commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -4967,6 +4967,24 @@ class TypeCheckCDeclFunctionRequest
 };
 
+/// A request to emit performance hints
+class EmitPerformanceHints
+: public SimpleRequest<EmitPerformanceHints,
+                       evaluator::SideEffect(SourceFile *),
+                       RequestFlags::Cached> {
```
Thank you for making this a separate request ;)

DougGregor commented on include/swift/Basic/Features.def:
```diff
@@ -553,6 +553,9 @@ EXPERIMENTAL_FEATURE(BorrowAndMutateAccessors, false)
 SUPPRESSIBLE_EXPERIMENTAL_FEATURE(InlineAlways, false)
 
+/// Emit experimental perfomance hints for low-level (embedded) Swift code
+EXPERIMENTAL_FEATURE(SwiftPerformanceHints, true)
```
You shouldn't need an experimental feature it. Performance hints are opt-in warnings. If you add the overall `PerformanceHints` warning group, then `-Wwarning PerformanceHints` or `-Werror PerformanceHints` will Just Work™️ to enable them as warnings or errors.

artemcm replied:
Removed, thanks!

DougGregor commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -8935,5 +8935,16 @@ ERROR(attr_inline_always_requires_inlinable,none,
+// MARK: Swift Performance hints
+//===----------------------------------------------------------------------===//
+
+GROUPED_WARNING(perf_hint_function_returns_array,PerfHintReturnTypeImplicitCopy,DefaultEscalate,
```
Instead of `DefaultEscalate`, I recommend `DefaultIgnore` to make these warnings be ignored by default.

DougGregor commented on include/swift/Option/Options.td:
```diff
@@ -929,6 +929,12 @@ def Wwarning : Separate<["-"], "Wwarning">,
   MetaVarName<"<diagnostic_group>">,
   HelpText<"Treat this warning group as warning">;
 
+def Wsuppress : Separate<["-"], "Wsuppress">,
+  Group<warning_treating_Group>,
+  Flags<[FrontendOption, HelpHidden]>,
+  MetaVarName<"<diagnostic_group>">,
+  HelpText<"Suppress this warning group">;
+
```
This was a big discussion in the review cycle of https://github.com/swiftlang/swift-evolution/blob/main/proposals/0443-warning-control-flags.md and it was intentionally not added to that proposal.

artemcm replied:
Removed, thanks for the pointer. `DefaultIgnore` fits better here.

DougGregor commented on lib/Sema/TypeChecker.cpp:
```diff
@@ -384,6 +384,10 @@ TypeCheckPrimaryFileRequest::evaluate(Evaluator &eval, SourceFile *SF) const {
 
+  // Opt-in performance hint diagnostics for performance-critical code.
+  if (Ctx.LangOpts.hasFeature(Feature::SwiftPerformanceHints))
+    evaluateOrDefault(Ctx.evaluator, EmitPerformanceHints{SF}, {});
```
FWIW, it's possible to check if specific warnings that are `DefaultIgnore` have been enabled. That would work here instead of having a separate experimental feature.

artemcm replied:
Implemented this instead as a check for whether any of the relevant default-ignored diagnostics are enabled in the current compilation.

DougGregor commented:
To be blunt, I don't really like the technical approach here, for a couple of reasons:
* We already have a mechanism for opt-in warnings (`DefaultIgnore` flag + diagnostic groups), and this should be based on that rather than adding an experimental feature (this isn't a language feature). 
* Adding a global `-Wsuppress <group name>` is a very big hammer and something that we need to socialize and discuss. It also feels like the wrong level of granularity for suppressing these issues: the warnings should be opt-in for a module (you get that with my suggestion from the previous bullet), so we don't also need some way to make them opt-out again at a module level.
* Using a performance-specific attribute (`@performanceOverride`) for warning suppression also feels like the wrong granularity, albeit in a different manner: it can be reasonable to suppress warnings within a region of code, but the performance hints aren't special in this way. Whether this should be an attribute or a file-level thing, I don't know, but it should apply to all warning groups.

DougGregor reviewed (approved):
This looks great, thank you!

artemcm commented:
@swift-ci smoke test

artemcm commented:
@swift-ci smoke test

artemcm commented on lib/Option/features.json:
```diff
@@ -53,6 +53,9 @@
        "name": "internal-import-bridging-header"
+    },
+    {
+      "name": "performance-hints"
```
Removed.
